File: services/chat-api/src/lib/ai/embedding.ts

/**
 * Embedding Generation Utility
 * Switched to Mistral Embeddings as Zeabur/Gemini Tokyo node lacks embedding support
 */

export async function generateEmbedding(text: string, inputType?: 'query' | 'document'): Promise<number[]> {
  const apiKey = process.env.VOYAGE_API_KEY;

  if (!apiKey) {
    console.warn('VOYAGE_API_KEY not found, using fallback embedding');
    return fallbackEmbedding(text);
  }

  try {
    const response = await fetch('https://api.voyageai.com/v1/embeddings', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${apiKey}`
      },
      body: JSON.stringify({
        model: 'voyage-4',
        input: [text],
        input_type: inputType
      })
    });

    if (!response.ok) {
      const errorData = await response.json().catch(() => ({ error: response.statusText }));
      console.error('Voyage Embedding API Error:', response.status, errorData);
      throw new Error(`Embedding API error: ${response.status}`);
    }

    const data: any = await response.json();
    if (!data.data?.[0]?.embedding) {
      console.error('Invalid embedding response format:', data);
      return fallbackEmbedding(text);
    }

    return data.data[0].embedding;
  } catch (error: any) {
    if (error.message?.includes('429')) {
      throw error;
    }
    console.error('Error generating embedding:', error);
    return fallbackEmbedding(text);
  }
}


function fallbackEmbedding(text: string): number[] {
  // Simple deterministic fallback for testing
  const dim = 1024; // Mistral embed is 1024 dim
  const result = new Array(dim).fill(0);
  for (let i = 0; i < text.length && i < dim; i++) {
    result[i] = text.charCodeAt(i) / 255;
  }
  return result;
}

File: services/vector-search-rs/src/embedding.rs
use reqwest::Client;
use serde_json::json;
use std::env;
use anyhow::{Result, anyhow};

// Voyage AI API endpoint
const EMBEDDING_API_URL: &str = "https://api.voyageai.com/v1/embeddings";

#[derive(serde::Deserialize)]
struct EmbeddingResponse {
    data: Vec<EmbeddingData>,
}

#[derive(serde::Deserialize)]
struct EmbeddingData {
    embedding: Vec<f32>,
}

pub async fn generate_embedding(text: &str, input_type: Option<&str>) -> Result<Vec<f32>> {
    let api_key = env::var("VOYAGE_API_KEY")
        .map_err(|_| anyhow!("VOYAGE_API_KEY not set"))?;
    
    let client = Client::new();
    
    let response = client.post(EMBEDDING_API_URL)
        .bearer_auth(&api_key)
        .json(&json!({
            "model": "voyage-4",
            "input": [text],
            "input_type": input_type
        }))
        .send()
        .await?;

    if !response.status().is_success() {
        let error_text = response.text().await?;
        return Err(anyhow!("Voyage AI API error: {}", error_text));
    }

    let result: EmbeddingResponse = response.json().await?;
    
    result.data.first()
        .map(|d| d.embedding.clone())
        .ok_or_else(|| anyhow!("No embedding returned from Voyage AI"))
}

